akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickProjectionImpl.scala

```diff
@@ -83,163 +77,32 @@ private[projection] class SlickProjectionImpl[Offset, Envelope, P <: JdbcProfile
-      try {
-        handlerFunc(env)
-      } catch {
-        case NonFatal(err) =>
-          handler.onFailure(env, err) match {
-
-            case Fail =>
-              logger.warning(
-                template =
-                  "An exception was thrown from inside the handler function while processing envelope with offset [{}]. Projection will stop as defined by recovery strategy.",
-                sourceProvider.extractOffset(env),
-                err.getClass.getSimpleName,
-                err.getMessage)
-              throw err
-
-            case Skip =>
-              logger.warning(
-                template =
-                  "An exception was thrown from inside the handler function while processing envelope with offset [{}]. Envelope will be skipped as defined by recovery strategy. Exception {}({})",
-                sourceProvider.extractOffset(env),
-                err.getClass.getSimpleName,
-                err.getMessage)
-              slick.dbio.DBIO.successful(Done)
+    implicit val dispatcher = systemProvider.classicSystem.dispatcher
 
-            case _: RetryAndSkip =>
-              logger.warning(
-                template =
-                  "An exception was thrown from inside the handler function while processing envelope with offset [{}]. This won't be retried. Envelope will be skipped as defined by recovery strategy. Exception {}({})",
-                sourceProvider.extractOffset(env),
-                err.getClass.getSimpleName,
-                err.getMessage)
-              slick.dbio.DBIO.successful(Done)
-
-            case _: RetryAndFail =>
-              logger.warning(
-                template =
-                  "An exception was thrown from inside the handler function while processing envelope with offset [{}]. This won't be retried. Projection will stop as defined by recovery strategy.",
-                sourceProvider.extractOffset(env),
-                err.getClass.getSimpleName,
-                err.getMessage)
-              throw err
-          }
-      }
-    }
+    def applyUserRecovery(envelope: Envelope)(futureCallback: () => Future[Done]): Future[Done] =
+      HandlerRecoveryImpl.applyUserRecovery[Offset, Envelope](handler, envelope, sourceProvider, logger, futureCallback)
 
     def processEnvelopeAndStoreOffsetInSameTransaction(env: Envelope): Future[Done] = {
       // run user function and offset storage on the same transaction
       // any side-effect in user function is at-least-once
       val txDBIO =
         offsetStore
           .saveOffset(projectionId, sourceProvider.extractOffset(env))
-          .flatMap(_ => aroundUserHandler(env)(handler.process))
+          .flatMap(_ => handler.process(env))
           .transactionally
 
       applyUserRecovery(env) { () =>
         databaseConfig.db.run(txDBIO).map(_ => Done)
       }
     }
 
-    def processEnvelopeAndStoreOffsetInSeparateTransactions(env: Envelope): Future[Done] = {
```

Comment:
removed this, and using future composition instead, see the `AtLeastOnce(1, _)` case below. Strange if offsetStore fails, then we would retry and run the handler transaction again.